File: app/packages/payy-client/src/bridge.ts
import { keccak_256 } from '@noble/hashes/sha3'
import { PayyClientError } from './errors'
import type {
  Address,
  B256,
  Hex,
  PayyMerklePath,
  PayyEvmReadClient,
  TxnData
} from './types'
import {
  assertHex,
  bytesToHex,
  encodeUtf8,
  hexToBytes,
  zeroHash
} from './utils'

const WORD_BYTES = 32

type WordArray3 = readonly [B256, B256, B256]
type WordArray4 = readonly [B256, B256, B256, B256]
type WordArray5 = readonly [B256, B256, B256, B256, B256]

export class BridgeClient {
  private readonly privacyBridge: Address
  private readonly readClient: PayyEvmReadClient

  constructor(privacyBridge: Address, readClient: PayyEvmReadClient) {
    this.privacyBridge = privacyBridge
    this.readClient = readClient
  }

  async getRoot(): Promise<B256> {
    return wordAt(await this.read(this.encodeGetRootCall()), 0)
  }

  async getMerklePath(commitment: B256): Promise<PayyMerklePath> {
    return decodeMerklePath(
      await this.read(this.encodeGetMerklePathCall(commitment))
    )
  }

  async elementExists(element: B256): Promise<boolean> {
    const word = wordAt(
      await this.read(this.encodeElementExistsCall(element)),
      0
    )
    if (word === zeroHash()) {
      return false
    }
    if (word === wordFromBigInt(1n)) {
      return true
    }
    throw malformedReturn('elementExists bool')
  }

  async getTxnHashByNonceHash(nonceHash: B256): Promise<B256 | null> {
    const value = wordAt(
      await this.read(this.encodeGetTxnHashByNonceHashCall(nonceHash)),
      0
    )
    return value === zeroHash() ? null : value
  }

  async getTxnHashByCommitment(commitment: B256): Promise<B256 | null> {
    const value = wordAt(
      await this.read(this.encodeGetTxnHashByCommitmentCall(commitment)),
      0
    )
    return value === zeroHash() ? null : value
  }

  async getTxnData(txnHash: B256): Promise<TxnData> {
    return decodeTxnData(await this.read(this.encodeGetTxnDataCall(txnHash)))
  }

  computeTxHash(
    verificationKeyHash: B256,
    proof: Hex,
    publicInputs: readonly B256[]
  ): B256 {
    return keccakHex(
      encodeBridgeProofParams(verificationKeyHash, proof, publicInputs)
    )
  }

  async getChainPublicKey(): Promise<readonly [B256, B256]> {
    const response = await this.read(this.encodeGetChainPublicKeyCall())
    return [wordAt(response, 0), wordAt(response, 1)]
  }

  encodeGetRootCall(): Hex {
    return encodeCall('getRoot()', [])
  }

  encodeGetMerklePathCall(commitment: B256): Hex {
    return encodeCall('getMerklePath(bytes32)', [word(commitment)])
  }

  encodeElementExistsCall(element: B256): Hex {
    return encodeCall('elementExists(bytes32)', [word(element)])
  }

  encodeGetTxnHashByNonceHashCall(nonceHash: B256): Hex {
    return encodeCall('getTxnHashByNonceHash(bytes32)', [word(nonceHash)])
  }

  encodeGetTxnHashByCommitmentCall(commitment: B256): Hex {
    return encodeCall('getTxnHashByCommitment(bytes32)', [word(commitment)])
  }

  encodeGetTxnDataCall(txnHash: B256): Hex {
    return encodeCall('getTxnData(bytes32)', [word(txnHash)])
  }

  encodeComputeTxHashCall(
    verificationKeyHash: B256,
    proof: Hex,
    publicInputs: readonly B256[]
  ): Hex {
    return encodeCall('computeTxHash(bytes32,bytes,bytes32[])', [
      encodeBridgeProofParams(verificationKeyHash, proof, publicInputs)
    ])
  }

  encodeGetChainPublicKeyCall(): Hex {
    return encodeCall('getChainPublicKey()', [])
  }

  encodeMintCall(
    verificationKeyHash: B256,
    proof: Hex,
    publicInputs: readonly B256[],
    userEncryptedKey: WordArray4 = zeroWordArray4()
  ): Hex {
    return encodeCall('mint(bytes32,bytes,bytes32[],bytes32[4])', [
      encodeProofCallParams(verificationKeyHash, proof, publicInputs, [
        ...userEncryptedKey
      ])
    ])
  }

  encodeBurnCall(
    verificationKeyHash: B256,
    proof: Hex,
    publicInputs: readonly B256[],
    userEncryptedKey: WordArray4 = zeroWordArray4()
  ): Hex {
    return encodeCall('burn(bytes32,bytes,bytes32[],bytes32[4])', [
      encodeProofCallParams(verificationKeyHash, proof, publicInputs, [
        ...userEncryptedKey
      ])
    ])
  }

  encodeTransferCall(
    verificationKeyHash: B256,
    proof: Hex,
    publicInputs: readonly B256[],
    bridgeMemo: B256,
    userEncryptedKey: WordArray4 = zeroWordArray4(),
    recipientEncryptedKey: WordArray4 = zeroWordArray4()
  ): Hex {
    return encodeCall(
      'transfer(bytes32,bytes,bytes32[],bytes32[4],bytes32[4],bytes32)',
      [
        encodeProofCallParams(verificationKeyHash, proof, publicInputs, [
          ...userEncryptedKey,
          ...recipientEncryptedKey,
          bridgeMemo
        ])
      ]
    )
  }

  private async read(data: Hex): Promise<Hex> {
    return this.readClient.readContract({
      to: this.privacyBridge,
      data
    })
  }
}

export function externalTransferTopic(): B256 {
  return eventSignatureTopic('ExternalTransfer(bytes6,bytes32)')
}

export function eventSignatureTopic(signature: string): B256 {
  return keccakHex(bytesToHex(encodeUtf8(signature)))
}

function encodeBridgeProofParams(
  verificationKeyHash: B256,
  proof: Hex,
  publicInputs: readonly B256[]
): Hex {
  return encodeProofCallParams(verificationKeyHash, proof, publicInputs, [])
}

function encodeProofCallParams(
  verificationKeyHash: B256,
  proof: Hex,
  publicInputs: readonly B256[],
  trailingStaticWords: readonly B256[]
): Hex {
  const proofTail = dynamicBytes(proof)
  const publicInputsTail = dynamicWords(publicInputs)
  const headWords = 3 + trailingStaticWords.length
  const proofOffset = BigInt(headWords * WORD_BYTES)
  const publicInputsOffset = proofOffset + BigInt(hexByteLength(proofTail))
  return concatHex([
    word(verificationKeyHash),
    wordFromBigInt(proofOffset),
    wordFromBigInt(publicInputsOffset),
    ...trailingStaticWords.map(word),
    proofTail,
    publicInputsTail
  ])
}

function encodeCall(signature: string, params: readonly Hex[]): Hex {
  return concatHex([selector(signature), ...params])
}

function selector(signature: string): Hex {
  return `0x${keccakHex(bytesToHex(encodeUtf8(signature))).slice(2, 10)}` as Hex
}

function decodeTxnData(response: Hex): TxnData {
  let index = 0
  const verificationKeyHash = wordAt(response, index)
  index += 1
  const senderEncryptedNote = wordArray5(response, index)
  index += 5
  const recipientEncryptedNote = wordArray5(response, index)
  index += 5
  const senderChainEncryptedKey = wordArray3(response, index)
  index += 3
  const recipientChainEncryptedKey = wordArray3(response, index)
  index += 3
  const userEncryptedKey = wordArray4(response, index)
  index += 4
  const recipientEncryptedKey = wordArray4(response, index)
  index += 4
  const memo = wordAt(response, index)
  return {
    verificationKeyHash,
    senderEncryptedNote,
    recipientEncryptedNote,
    senderChainEncryptedKey,
    recipientChainEncryptedKey,
    userEncryptedKey,
    recipientEncryptedKey,
    memo
  }
}

function decodeMerklePath(response: Hex): PayyMerklePath {
  const root = wordAt(response, 0)
  const offset = Number(wordBigInt(wordAt(response, 1)))
  if (offset % WORD_BYTES !== 0) {
    throw malformedReturn('getMerklePath offset')
  }
  const length = Number(wordBigInt(wordAtByteOffset(response, offset)))
  const start = offset + WORD_BYTES
  const out: B256[] = []
  for (let i = 0; i < length; i += 1) {
    out.push(wordAtByteOffset(response, start + i * WORD_BYTES))
  }
  return { root, siblings: out }
}

function wordArray3(response: Hex, index: number): WordArray3 {
  return [
    wordAt(response, index),
    wordAt(response, index + 1),
    wordAt(response, index + 2)
  ]
}

function wordArray4(response: Hex, index: number): WordArray4 {
  return [
    wordAt(response, index),
    wordAt(response, index + 1),
    wordAt(response, index + 2),
    wordAt(response, index + 3)
  ]
}

function wordArray5(response: Hex, index: number): WordArray5 {
  return [
    wordAt(response, index),
    wordAt(response, index + 1),
    wordAt(response, index + 2),
    wordAt(response, index + 3),
    wordAt(response, index + 4)
  ]
}

function wordAt(response: Hex, index: number): B256 {
  return wordAtByteOffset(response, index * WORD_BYTES)
}

function wordAtByteOffset(response: Hex, offset: number): B256 {
  assertHex(response)
  const start = 2 + offset * 2
  const end = start + WORD_BYTES * 2
  if (response.length < end) {
    throw malformedReturn('short contract return')
  }
  return `0x${response.slice(start, end)}` as B256
}

function word(value: B256): Hex {
  assertHex(value, 32)
  return value
}

function wordFromBigInt(value: bigint): B256 {
  return `0x${value.toString(16).padStart(64, '0')}` as B256
}

function wordBigInt(value: B256): bigint {
  return BigInt(value)
}

function dynamicBytes(value: Hex): Hex {
  const bytes = hexToBytes(value)
  return concatHex([
    wordFromBigInt(BigInt(bytes.length)),
    bytesToHex(padRight(bytes, WORD_BYTES))
  ])
}

function dynamicWords(values: readonly B256[]): Hex {
  return concatHex([
    wordFromBigInt(BigInt(values.length)),
    ...values.map((value) => word(value))
  ])
}

function padRight(bytes: Uint8Array, multiple: number): Uint8Array {
  const paddedLength = Math.ceil(bytes.length / multiple) * multiple
  const out = new Uint8Array(paddedLength)
  out.set(bytes)
  return out
}

function concatHex(parts: readonly Hex[]): Hex {
  return `0x${parts.map((part) => part.slice(2)).join('')}` as Hex
}

function hexByteLength(value: Hex): number {
  assertHex(value)
  return (value.length - 2) / 2
}

function keccakHex(value: Hex): B256 {
  return bytesToHex(keccak_256(hexToBytes(value)))
}

function zeroWordArray4(): WordArray4 {
  return [zeroHash(), zeroHash(), zeroHash(), zeroHash()]
}

function malformedReturn(detail: string): PayyClientError {
  return new PayyClientError(
    'contract_return_malformed',
    `malformed PrivacyBridge return: ${detail}`
  )
}

File: pkg/payy-evm-client/src/bridge.rs
// lint-long-file-override allow-max-lines=300
mod abi;

use std::sync::Arc;

use payy_evm_client_interface::{B256, Bytes, PayyEvmCallRequest, Result, TxnData};

use crate::client::ClientInner;

/// Raw PrivacyBridge read surface.
#[derive(Clone)]
pub struct BridgeClient {
    inner: Arc<ClientInner>,
}

/// Root and sibling witness returned by `getMerklePath`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MerklePathResult {
    /// Root paired with this path by the bridge read.
    pub root: B256,
    /// Sibling words for the requested commitment.
    pub siblings: Vec<B256>,
}

impl BridgeClient {
    pub(crate) fn new(inner: Arc<ClientInner>) -> Self {
        Self { inner }
    }

    #[must_use]
    pub(crate) fn encode_get_root_call() -> Bytes {
        abi::encode_get_root_call()
    }

    #[must_use]
    pub(crate) fn encode_get_merkle_path_call(commitment: B256) -> Bytes {
        abi::encode_get_merkle_path_call(commitment)
    }

    #[must_use]
    pub(crate) fn encode_element_exists_call(element: B256) -> Bytes {
        abi::encode_element_exists_call(element)
    }

    #[must_use]
    pub(crate) fn encode_get_txn_hash_by_nonce_hash_call(nonce_hash: B256) -> Bytes {
        abi::encode_get_txn_hash_by_nonce_hash_call(nonce_hash)
    }

    #[must_use]
    pub(crate) fn encode_get_txn_hash_by_commitment_call(commitment: B256) -> Bytes {
        abi::encode_get_txn_hash_by_commitment_call(commitment)
    }

    #[must_use]
    pub(crate) fn encode_get_txn_data_call(txn_hash: B256) -> Bytes {
        abi::encode_get_txn_data_call(txn_hash)
    }

    #[must_use]
    pub(crate) fn encode_get_chain_public_key_call() -> Bytes {
        abi::encode_get_chain_public_key_call()
    }

    /// Read current root.
    pub async fn get_root(&self) -> Result<B256> {
        let bytes = self.read(Self::encode_get_root_call()).await?;
        abi::decode_b256("getRoot", &bytes)
    }

    /// Read Merkle path and the root it was generated against.
    pub async fn get_merkle_path(&self, commitment: B256) -> Result<MerklePathResult> {
        let bytes = self
            .read(Self::encode_get_merkle_path_call(commitment))
            .await?;
        let (root, siblings) = abi::decode_merkle_path(&bytes)?;
        Ok(MerklePathResult { root, siblings })
    }

    /// Check element existence.
    pub async fn element_exists(&self, element: B256) -> Result<bool> {
        let bytes = self.read(Self::encode_element_exists_call(element)).await?;
        abi::decode_bool("elementExists", &bytes)
    }

    /// Resolve txn hash by nonce hash.
    pub async fn get_txn_hash_by_nonce_hash(&self, nonce_hash: B256) -> Result<Option<B256>> {
        self.read_optional_b256(Self::encode_get_txn_hash_by_nonce_hash_call(nonce_hash))
            .await
    }

    /// Resolve txn hash by commitment.
    pub async fn get_txn_hash_by_commitment(&self, commitment: B256) -> Result<Option<B256>> {
        self.read_optional_b256(Self::encode_get_txn_hash_by_commitment_call(commitment))
            .await
    }

    /// Resolve txn data.
    pub async fn get_txn_data(&self, txn_hash: B256) -> Result<TxnData> {
        let bytes = self.read(Self::encode_get_txn_data_call(txn_hash)).await?;
        abi::decode_txn_data(&bytes)
    }

    /// Compute bridge transaction hash.
    #[must_use]
    pub fn compute_tx_hash(
        &self,
        verification_key_hash: B256,
        proof: &[u8],
        public_inputs: &[B256],
    ) -> B256 {
        abi::compute_tx_hash(verification_key_hash, proof, public_inputs)
    }

    /// Read chain public key.
    pub async fn get_chain_public_key(&self) -> Result<(B256, B256)> {
        let bytes = self.read(Self::encode_get_chain_public_key_call()).await?;
        abi::decode_chain_public_key(&bytes)
    }

    #[must_use]
    pub(crate) fn encode_mint_call(
        verification_key_hash: B256,
        proof: &[u8],
        public_inputs: &[B256],
        user_encrypted_key: [B256; 4],
    ) -> Bytes {
        abi::encode_mint_call(
            verification_key_hash,
            proof,
            public_inputs,
            user_encrypted_key,
        )
    }

    #[must_use]
    pub(crate) fn encode_burn_call(
        verification_key_hash: B256,
        proof: &[u8],
        public_inputs: &[B256],
        user_encrypted_key: [B256; 4],
    ) -> Bytes {
        abi::encode_burn_call(
            verification_key_hash,
            proof,
            public_inputs,
            user_encrypted_key,
        )
    }

    #[must_use]
    pub(crate) fn encode_transfer_call(
        verification_key_hash: B256,
        proof: &[u8],
        public_inputs: &[B256],
        user_encrypted_key: [B256; 4],
        recipient_encrypted_key: [B256; 4],
        memo: B256,
    ) -> Bytes {
        abi::encode_transfer_call(
            verification_key_hash,
            proof,
            public_inputs,
            user_encrypted_key,
            recipient_encrypted_key,
            memo,
        )
    }

    async fn read(&self, data: Bytes) -> Result<Bytes> {
        self.inner.validate_read_chain().await?;
        self.inner
            .read_client
            .read_contract(PayyEvmCallRequest {
                to: self.inner.network.privacy_bridge,
                data,
                block_number: None,
            })
            .await
    }

    async fn read_optional_b256(&self, data: Bytes) -> Result<Option<B256>> {
        let bytes = self.read(data).await?;
        let value = abi::decode_b256("bytes32 lookup", &bytes)?;
        Ok((value != [0; 32]).then_some(value))
    }
}
